segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/MetadataStore.java

```diff
@@ -0,0 +1,717 @@
+    public MetadataStore(@NonNull Connector connector, @NonNull Executor executor) {
+        this.traceObjectId = String.format("MetadataStore[%d]", connector.containerMetadata.getContainerId());
+        this.connector = connector;
+        this.executor = executor;
+        this.pendingRequests = new HashMap<>();
+    }
+
+    //endregion
+
+    //region Create Segments
+
+    /**
+     * Creates a new Segment with given name.
+     *
+     * @param segmentName The case-sensitive Segment Name.
+     * @param attributes  The initial attributes for the StreamSegment, if any.
+     * @param timeout     Timeout for the operation.
+     * @return A CompletableFuture that, when completed normally, will indicate the Segment has been created.
+     * If the operation failed, this will contain the exception that caused the failure. Notable exceptions:
+     * <ul>
+     * <li>{@link StreamSegmentExistsException} If the Segment already exists.
+     * </ul>
+     */
+    CompletableFuture<Void> createSegment(String segmentName, Collection<AttributeUpdate> attributes, Duration timeout) {
+        long traceId = LoggerHelpers.traceEnterWithContext(log, traceObjectId, "createSegment", segmentName);
+        long segmentId = this.connector.containerMetadata.getStreamSegmentId(segmentName, true);
+        if (isValidSegmentId(segmentId)) {
+            // Quick fail: see if this is an active Segment, and if so, don't bother with anything else.
+            return Futures.failedFuture(new StreamSegmentExistsException(segmentName));
+        }
+
+        ArrayView segmentInfo = SegmentInfo.serialize(SegmentInfo.newSegment(segmentName, attributes));
+        CompletableFuture<Void> result = createSegment(segmentName, segmentInfo, new TimeoutTimer(timeout));
+        if (log.isTraceEnabled()) {
+            result.thenAccept(v -> LoggerHelpers.traceLeave(log, traceObjectId, "createSegment", traceId, segmentName));
+        }
+
+        return result;
+    }
+
+    /**
+     * Attempts to create the given Segment in the Metadata Store.
+     *
+     * @param segmentName The name of the Segment to create.
+     * @param segmentInfo An {@link ArrayView} containing the serialized Segment Info to store for this segment.
+     * @param timer       A TimeoutTimer for the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the Segment has been successfully created.
+     */
+    protected abstract CompletableFuture<Void> createSegment(String segmentName, ArrayView segmentInfo, TimeoutTimer timer);
+
+    //endregion
+
+    //region Delete Segments
+
+    /**
+     * Deletes a Segment and any associated information from the Metadata Store.
+     * Notes:
+     * - This method removes both the Segment and its Metadata Store entries.
+     * - {@link #clearSegmentInfo} only removes Metadata Store entries.
+     *
+     * This operation is made of multiple steps and is restart-able. If it was only able to execute partially before being
```

Comment:
Based on my understanding, "idempotent" means an operation can be called several times with the same result. In this case, calling delete after a successful call to delete (on the same segment) will result in an exception (Segment does not exist).

`restart-able` means if the operation was interrupted mid-way (due to a crash), then reinvoking it will pick up from where it left off and finish the job.

Reply:
Ok, got it.